src/librustc_typeck/check/expr.rs

```diff
@@ -548,7 +548,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     coerce.coerce(self, &cause, e, e_ty);
                 } else {
                     assert!(e_ty.is_unit());
-                    coerce.coerce_forced_unit(self, &cause, &mut |_| (), true);
+                    let ty = coerce.expected_ty();
+                    coerce.coerce_forced_unit(self, &cause, &mut |err| {
+                        let msg = "give it a value of the expected type";
+                        let label = destination.label
+                            .map(|l| format!(" {}", l.ident))
+                            .unwrap_or_else(String::new);
+                        let sugg = format!("break{} {}", label, match ty.sty {
+                            ty::Bool => "true",
+                            ty::Char => "'a'",
+                            ty::Int(_) | ty::Uint(_) => "42",
+                            ty::Float(_) => "3.14159",
+                            ty::Error | ty::Never => return,
```

Comment:
Possibly?
```suggestion
                            ty::Never => "loop {}",
                            ty::Error => return,
```
`panic!()` could also be a candidate.

Reply:
I've thought about that, but that suggestion is more likely than not going to make little sense to people. I'd like to see `!` being used in the wild for a while before deciding what to suggest for it in errors.

Reply:
Yeah that's a good point.

Reply:
https://github.com/rust-lang/rust/issues/63390